Implement HTMLTextArea.setSelectionRange

0X1A opened the pull request:
Fixes #9994

Correctly passes all tests in `tests/wpt/web-platform-tests/html/semantics/forms/textfieldselection/textfieldselection-setSelectionRange.html`

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/10007)

<!-- Reviewable:end -->

0X1A committed:
Implement HTMLTextArea.setSelectionRange

ecoal95 commented:
Nice, it looks good, thanks for doing this!

This needs a bit more of refactoring work to reuse the common logic between `htmlinputelement.rs` and `htmltextareaelement.rs`, and after that we can land it! :)

-S-awaiting-review +S-needs-code-changes 

---

Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[components/script/dom/htmltextareaelement.rs, line 37 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10007#-KCuh8O42Yy6yH99XC4s:-KCuh8O5AlqDAnihY5xj:-1203611775) ([raw file](https://github.com/servo/servo/blob/77ec22bd72acdff4a3d67301e18c1fd9add83ef0/components/script/dom/htmltextareaelement.rs#L37)):*
This should not exist. We should instead reuse the type in `htmlinputelement.rs`.

Since it's common to both of them, it seems moving it to `components/script/textinput.rs` is the best idea.

---

*[components/script/dom/htmltextareaelement.rs, line 127 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10007#-KCuhWnPrfllUJ9YwHaW:-KCuhWnPrfllUJ9YwHaX:-2106648236) ([raw file](https://github.com/servo/servo/blob/77ec22bd72acdff4a3d67301e18c1fd9add83ef0/components/script/dom/htmltextareaelement.rs#L127)):*
Since this method is exactly the same than the one is `htmlinputelement.rs`, we should probably re-use it.

The easiest way fot it, given it mostly uses `text_input`, should be moving it so it's a method of `TextInput`, with a signature like:

```rust
fn set_selection_range(&mut self, start: u32, end: u32, direction: &SelectionDirection)
```

So we only need to update the `selection_direction` member of `HTMLInputElement` or `HTMLTextareaElement` after calling it.

In fact, it seems even logic to move the `selection_direction` field to `TextInput` directly, and implement most of this logic there.

---

*[components/script/dom/htmltextareaelement.rs, line 275 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10007#-KCuiy5Ie1-MV1jprObY:-KCuiy5JSh49dKe144dS:69187366) ([raw file](https://github.com/servo/servo/blob/77ec22bd72acdff4a3d67301e18c1fd9add83ef0/components/script/dom/htmltextareaelement.rs#L275)):*
Same thing, this logic should probably be reused as part of `TextInput`, so this method would become something like:

```
fn SelectionStart(&self) -> u32 {
    self.textinput.borrow().selection_start()
}
```

---

*[tests/wpt/metadata/html/semantics/forms/textfieldselection/textfieldselection-setSelectionRange.html.ini, line 4 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10007#-KCujIiOAabMFUrrvnoe:-KCujIiOAabMFUrrvnof:1329047224) ([raw file](https://github.com/servo/servo/blob/77ec22bd72acdff4a3d67301e18c1fd9add83ef0/tests/wpt/metadata/html/semantics/forms/textfieldselection/textfieldselection-setSelectionRange.html.ini#L4)):*
nit: There's no need to explicitely specify "PASS" here, since it's the default.

Since there are no remaining tests failing here, this file should be deleted entirely.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/10007#-:-KCujcdHzN6gkyh8OTRL:-591016538)*
<!-- Sent from Reviewable.io -->

jdm commented:
```
80.10s$ ./mach test-tidy --no-progress
./components/script/dom/htmltextareaelement.rs:244: link to WHATWG may break in the future, use this format instead: https://html.spec.whatwg.org/multipage/#dom-textarea

./components/script/dom/htmltextareaelement.rs:249: link to WHATWG may break in the future, use this format instead: https://html.spec.whatwg.org/multipage/#dom-textarea

./components/script/dom/htmltextareaelement.rs:258: link to WHATWG may break in the future, use this format instead: https://html.spec.whatwg.org/multipage/#dom-textarea

./components/script/dom/htmltextareaelement.rs:263: link to WHATWG may break in the future, use this format instead: https://html.spec.whatwg.org/multipage/#dom-textarea

./components/script/dom/htmltextareaelement.rs:269: link to WHATWG may break in the future, use this format instead: https://html.spec.whatwg.org/multipage/#dom-textarea

./components/script/dom/htmltextareaelement.rs:274: link to WHATWG may break in the future, use this format instead: https://html.spec.whatwg.org/multipage/#dom-textarea

./components/script/dom/htmltextareaelement.rs:287: link to WHATWG may break in the future, use this format instead: https://html.spec.whatwg.org/multipage/#dom-textarea
```

0X1A committed:
Fix WHATWG links

0X1A commented:
Review status: 2 of 3 files reviewed at latest revision, 4 unresolved discussions.

---

*[components/script/dom/htmltextareaelement.rs, line 37 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10007#-KCuh8O42Yy6yH99XC4s:-KCuqqpFrcmix3Em7Nu9:-2026634333) ([raw file](https://github.com/servo/servo/blob/df0150dbb99cc467a88cf70eeceb3ad908a44cf9/components/script/dom/htmltextareaelement.rs#L37)):*
Sounds good, moving these methods to `textinput.rs` for re-usability.

---

*[tests/wpt/metadata/html/semantics/forms/textfieldselection/textfieldselection-setSelectionRange.html.ini, line 4 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10007#-KCujIiOAabMFUrrvnoe:-KCur65KG3BM9sRHBVgm:-69688507) ([raw file](https://github.com/servo/servo/blob/df0150dbb99cc467a88cf70eeceb3ad908a44cf9/tests/wpt/metadata/html/semantics/forms/textfieldselection/textfieldselection-setSelectionRange.html.ini#L4)):*
Wasn't aware these tests defaulted to PASS, this file will be deleted

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/10007)*
<!-- Sent from Reviewable.io -->

0X1A committed:
Move SelectionDirection, add re-usable methods

0X1A committed:
Use re-usable methods from TextInput

	- Use re-usable methods for implementing `setSelectionRange`
	  from `TextInput`
	- Move SelectionDirection from `htmlinputelement.rs` to
	  `textinput.rs`

0X1A commented:
@ecoal95 Since `SelectionDirection` is being added to `TextInput`, should its value be changed in unit tests? I assume it should default to `SelectionDirection::None` in all the calls to `new`

ecoal95 commented:
Looks good! We still have a few unnecessary allocations, and the unit test changes. Regarding those, sure! `SelectionDirection::None` seems like the correct default value.

You seem to have done a merge which has introduced merge conflicts with master, please rebase your branch so this can merge with master. If you need some help with it feel free to say it over here or over the `#servo` channel on irc :)

-S-awaiting-review +S-needs-code-changes +S-needs-rebase 

---

Reviewed 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[components/script/textinput.rs, line 552 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10007#-KCvpKBjDc1Tar0kDBre:-KCvpKBjDc1Tar0kDBrf:1202298330) ([raw file](https://github.com/servo/servo/blob/1e4c736e193ff84044b24b9cbf692051f31dfd87/components/script/textinput.rs#L552)):*
I'd prefer to keep `DOMString`s out of this file, or at least out of most of the functions, since it's not necessary. It'll also prevent some needless allocations.

---

*[components/script/dom/htmlinputelement.rs, line 462 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10007#-KCvpfHlZj5CV3MoXCss:-KCvpfHm0hFB9ArjsXdg:182889001) ([raw file](https://github.com/servo/servo/blob/1e4c736e193ff84044b24b9cbf692051f31dfd87/components/script/dom/htmlinputelement.rs#L462)):*
This is one of the allocations we could prevent, for example :)

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/10007#-:-KCvpwCAGTpX1Ue-XAvD:2030801106)*
<!-- Sent from Reviewable.io -->

bors-servo commented:
:umbrella: The latest upstream changes (presumably #10013) made this pull request unmergeable. Please resolve the merge conflicts.

0X1A committed:
Remove DOMString from `set_selection_range`

0X1A committed:
Update TextInput unit test

0X1A commented:
@ecoal95 *Think* I did a proper rebase, how's this looking?

ecoal95 commented:
The rebase it's ok, except for the test file that came back :P 

I think that the removed `interfaces.html.ini` lines came back too ;)

It looks good! I left a few other nits, but they are mostly minor, and I think they can be adressed later in a follow-up if you want :)

-S-awaiting-review +S-needs-code-changes 

---

Reviewed 5 of 5 files at r4.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[components/script/textinput.rs, line 552 \[r4\]](https://reviewable.io:443/reviews/servo/servo/10007#-KD5cbeYH6LsiQnFUqCX:-KD5cbeYH6LsiQnFUqCY:-730879431) ([raw file](https://github.com/servo/servo/blob/ec31ee883bafdb27d964d599e7f954d43133cdfa/components/script/textinput.rs#L552)):*
I was thinking about making the `direction` parameter a `SelectionDirection`, but this can work too if the selection is updated appropiately and it isn't too much bloat (since I think most of the time this is called it's done for the same selection direction). 

We can always do later a `set_selection_range_and_direction` method.

---

*[components/script/textinput.rs, line 569 \[r4\]](https://reviewable.io:443/reviews/servo/servo/10007#-KD5dk4qDw-DYDRD4crO:-KD5dk4rZZL0YN5rbGZi:-257988373) ([raw file](https://github.com/servo/servo/blob/ec31ee883bafdb27d964d599e7f954d43133cdfa/components/script/textinput.rs#L569)):*
I don't quite like having the `selection_direction` field made public just because this setter takes a `DOMString` instead of a `SelectionDirection`.

It's not a huge issue, since it's not used a lot, but we could `impl<'a> From<&'a str> for SelectionDirection {...}`, or similar, to leverage it.

---

*[tests/wpt/metadata/html/semantics/forms/textfieldselection/textfieldselection-setSelectionRange.html.ini, line 1 \[r4\]](https://reviewable.io:443/reviews/servo/servo/10007#-KCwhSIciHhdgjl5wE3K:-KCwhSIciHhdgjl5wE3L:272643355) ([raw file](https://github.com/servo/servo/blob/ec31ee883bafdb27d964d599e7f954d43133cdfa/tests/wpt/metadata/html/semantics/forms/textfieldselection/textfieldselection-setSelectionRange.html.ini#L1)):*
Whoops, they came back! :P

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/10007#-:-KD5eecJEFcZdf1vqg2J:2077939214)*
<!-- Sent from Reviewable.io -->

0X1A commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[components/script/textinput.rs, line 569 \[r4\]](https://reviewable.io:443/reviews/servo/servo/10007#-KD5dk4qDw-DYDRD4crO:-KD5hNVGVProRYW7PwyL:1987507933) ([raw file](https://github.com/servo/servo/blob/ec31ee883bafdb27d964d599e7f954d43133cdfa/components/script/textinput.rs#L569)):*
Was thinking the same thing of using the trait `From`. The reason I left these methods in was because they were used in both `htmlinputelement.rs` and `htmltextareaelemnt.rs` and found it'd be very redundant to have the same code in both files. I believe a `From` impl for `SelectionDirection` should get rid of having to use these methods.

---

*[tests/wpt/metadata/html/semantics/forms/textfieldselection/textfieldselection-setSelectionRange.html.ini, line 1 \[r4\]](https://reviewable.io:443/reviews/servo/servo/10007#-KCwhSIciHhdgjl5wE3K:-KD5iI7AupmuUo6P80Yx:1298714393) ([raw file](https://github.com/servo/servo/blob/ec31ee883bafdb27d964d599e7f954d43133cdfa/tests/wpt/metadata/html/semantics/forms/textfieldselection/textfieldselection-setSelectionRange.html.ini#L1)):*
Must've missed this in the rebase

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/10007)*
<!-- Sent from Reviewable.io -->

0X1A committed:
Use `From` for DOMString <=> SelectionDirection

	- Use `From` trait for conversion from SelectionDirection to
	  DOMString and vice versa.

0X1A committed:
Remove unneeded ini for `setSelectionRange`

emilio commented:
Hey, this looks great! Thanks for your work! :)

Sorry for the delay, I hate github not pinging when new commits are pushed.

@bors-servo: r+

-S-awaiting-review +S-awaiting-merge 

---

Review status: 2 of 6 files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/10007#-:-KDrw-QbDfMgsRU2hWYv:861542970)*
<!-- Sent from Reviewable.io -->

bors-servo commented:
:pushpin: Commit 332e6ca has been approved by `emilio`

<!-- @bors-servo r=emilio 332e6ca4cecb185d2200107ec91c41fb0a668d87 -->

bors-servo commented:
:hourglass: Testing commit 332e6ca with merge 56b4d3f...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2335)

emilio commented:
Upps, it's true, you borrow textinput mutably twice some times, so you'll have to address that, since there are a few tests crashing:

```
Tests with unexpected results:
  ▶ Unexpected subtest result in /html/dom/interfaces.html:
  └ PASS [expected FAIL] HTMLTextAreaElement interface: attribute selectionStart

  ▶ Unexpected subtest result in /html/dom/interfaces.html:
  └ PASS [expected FAIL] HTMLTextAreaElement interface: attribute selectionEnd

  ▶ Unexpected subtest result in /html/dom/interfaces.html:
  └ PASS [expected FAIL] HTMLTextAreaElement interface: attribute selectionDirection

  ▶ Unexpected subtest result in /html/dom/interfaces.html:
  └ PASS [expected FAIL] HTMLTextAreaElement interface: operation setSelectionRange(unsigned long,unsigned long,DOMString)

  ▶ Unexpected subtest result in /html/dom/interfaces.html:
  └ PASS [expected FAIL] HTMLTextAreaElement interface: document.createElement("textarea") must inherit property "selectionStart" with the proper type (27)

  ▶ Unexpected subtest result in /html/dom/interfaces.html:
  └ PASS [expected FAIL] HTMLTextAreaElement interface: document.createElement("textarea") must inherit property "selectionEnd" with the proper type (28)

  ▶ Unexpected subtest result in /html/dom/interfaces.html:
  └ PASS [expected FAIL] HTMLTextAreaElement interface: document.createElement("textarea") must inherit property "selectionDirection" with the proper type (29)

  ▶ Unexpected subtest result in /html/dom/interfaces.html:
  └ PASS [expected FAIL] HTMLTextAreaElement interface: document.createElement("textarea") must inherit property "setSelectionRange" with the proper type (32)

  ▶ Unexpected subtest result in /html/dom/interfaces.html:
  └ PASS [expected FAIL] HTMLTextAreaElement interface: calling setSelectionRange(unsigned long,unsigned long,DOMString) on document.createElement("textarea") with too few arguments must throw TypeError

  ▶ CRASH [expected OK] /html/semantics/forms/textfieldselection/textfieldselection-setRangeText.html
  │ 
  │ thread 'ScriptThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(0) }' panicked at 'DOMRefCell<T> already mutably borrowed', ../src/libcore/option.rs:704
  │ stack backtrace:
  │    1:        0x1061b3e58 - sys::backtrace::tracing::imp::write::h3153bab3099b48114Zu
  │    2:        0x1061b7a49 - panicking::default_handler::_$u7b$$u7b$closure$u7d$$u7d$::closure.43920
  │    3:        0x1061b7678 - panicking::default_handler::h1183b2a11c8682560zz
  │    4:        0x1061a2a46 - sys_common::unwind::begin_unwind_inner::h1f97bc9e4cfe0396LZt
  │    5:        0x1061a3a7e - sys_common::unwind::begin_unwind_fmt::hda8646f04aa67d8aRYt
  │    6:        0x1061b31f7 - rust_begin_unwind
  │    7:        0x1061dd650 - panicking::panic_fmt::h40f5ec0cdc3fc429FRL
  │    8:        0x1061e8935 - option::expect_failed::h56e104933d0889c7LpO
  │    9:        0x104ba1e7e - dom::htmlinputelement::HTMLInputElement.HTMLInputElementMethods::SetSelectionStart::hb411b730858a10c0HM3
  │   10:        0x104f9caf0 - dom::bindings::codegen::Bindings::HTMLInputElementBinding::set_selectionStart::hc29799c26c5054eeDdo
  │   11:        0x1058790aa - CallJitSetterOp
  │   12:        0x104f60a5d - dom::bindings::utils::call_setter::h3621e47e57f90dd8Sug
  │   13:        0x104f608df - dom::bindings::utils::generic_call::h2289bd9fb0388cfdcqg
  │   14:        0x1059e597b - _ZN2js6InvokeEP9JSContextN2JS8CallArgsENS_14MaybeConstructE
  │   15:        0x1059ddf7d - _ZN2js6InvokeEP9JSContextRKN2JS5ValueES5_jPS4_NS2_13MutableHandleIS3_EE
  │   16:        0x1059fbbae - _ZN2js20InvokeGetterOrSetterEP9JSContextP8JSObjectN2JS5ValueEjPS5_NS4_13MutableHandleIS5_EE
  │   17:        0x105a19937 - _ZN2js17NativeSetPropertyEP9JSContextN2JS6HandleIPNS_12NativeObjectEEENS3_IP8JSObjectEENS3_I4jsidEENS_13QualifiedBoolENS2_13MutableHandleINS2_5ValueEEERNS2_14ObjectOpResultE
  │   18:        0x1059f4a95 - _ZL9InterpretP9JSContextRN2js8RunStateE
  │   19:        0x1059eceb9 - _ZN2js9RunScriptEP9JSContextRNS_8RunStateE
  │   20:        0x1059e5bb7 - _ZN2js6InvokeEP9JSContextN2JS8CallArgsENS_14MaybeConstructE
  │   21:        0x105d1095c - _ZN2js9fun_applyEP9JSContextjPN2JS5ValueE
  │   22:        0x1059e597b - _ZN2js6InvokeEP9JSContextN2JS8CallArgsENS_14MaybeConstructE
  │   23:        0x1059f64d9 - _ZL9InterpretP9JSContextRN2js8RunStateE
  │   24:        0x1059eceb9 - _ZN2js9RunScriptEP9JSContextRNS_8RunStateE
  │   25:        0x1059e5bb7 - _ZN2js6InvokeEP9JSContextN2JS8CallArgsENS_14MaybeConstructE
  │   26:        0x1059f64d9 - _ZL9InterpretP9JSContextRN2js8RunStateE
  │   27:        0x1059eceb9 - _ZN2js9RunScriptEP9JSContextRNS_8RunStateE
  │   28:        0x1059fbd36 - _ZN2js13ExecuteKernelEP9JSContextN2JS6HandleIP8JSScriptEER8JSObjectRKNS2_5ValueENS_11ExecuteTypeENS_16AbstractFramePtrEPS9_
  │   29:        0x1059fbe63 - _ZN2js7ExecuteEP9JSContextN2JS6HandleIP8JSScriptEER8JSObjectPNS2_5ValueE
  │   30:        0x105cdbf02 - _ZL8EvaluateP9JSContextN2JS6HandleIP8JSObjectEERKNS1_22ReadOnlyCompileOptionsERNS1_18SourceBufferHolderENS1_13MutableHandleINS1_5ValueEEE
  │   31:        0x105cdc06a - _ZN2JS8EvaluateEP9JSContextRKNS_22ReadOnlyCompileOptionsEPKDsmNS_13MutableHandleINS_5ValueEEE
  │   32:        0x104ff75fa - dom::htmlscriptelement::HTMLScriptElement::execute::hd985d34a6a844843Sr6
  │   33:        0x105024fa9 - dom::htmlscriptelement::HTMLScriptElement::prepare::he74a5354aaa7d643mh6
  │   34:        0x104bc6bb0 - tree_builder::rules::super..TreeBuilder<Handle, Sink>.TreeBuilderStep::step::h14576230894425074504
  │   35:        0x104bbe90f - tree_builder::TreeBuilder<Handle, Sink>.TokenSink::process_token::h7834235262675994692
  │   36:        0x104bbdef1 - tokenizer::Tokenizer<Sink>::process_token::h6160948021956107981
  │   37:        0x104bee34a - tokenizer::Tokenizer<Sink>::emit_current_tag::h15594130976656807786
  │   38:        0x104be86dd - tokenizer::Tokenizer<Sink>::step::h16952525606794884413
  │   39:        0x104be38c7 - tokenizer::Tokenizer<Sink>::run::h13044294324310735172
  │   40:        0x104bb8c88 - dom::servohtmlparser::ServoHTMLParser::parse_sync::h9952bd596f81f8e2cSf
  │   41:        0x104ff41bf - dom::document::Document::finish_load::h89ddd8470eb96f7eVLO
  │   42:        0x10502350d - dom::htmlscriptelement::ScriptContext.AsyncResponseListener::response_complete::h2b001ff6297d70d3Sf6
  │   43:        0x105750dda - ResponseAction::process::h69a97ad5099cd1a9Ypf
  │   44:        0x105026bb0 - network_listener::ListenerRunnable<T>.Runnable::handler::h5208449697483038179
  │   45:        0x104ec99bb - script_thread::ScriptThread::handle_msg_from_script::h0b0c9a560280f108Xzt
  │   46:        0x104f2a09a - script_thread::ScriptThread::handle_msgs::_$u7b$$u7b$closure$u7d$$u7d$::closure.188824
  │   47:        0x104f12efe - script_thread::ScriptThread::handle_msgs::h45aab595807156d2Blt
  │   48:        0x104eb93c7 - sys_common::unwind::try::try_fn::h3279682059033863226
  │   49:        0x1061b318b - __rust_try
  │   50:        0x1061b3113 - sys_common::unwind::inner_try::hdf6affff6fc30ecdNWt
  │   51:        0x104eba81a - boxed::F.FnBox<A>::call_box::h15091847669718450595
  │   52:        0x1061b6b2c - sys::thread::Thread::new::thread_start::he33a0e5f2597fcdejOy
  │   53:     0x7fff8face059 - _pthread_body
  └   54:     0x7fff8facdfd6 - _pthread_start

  ▶ CRASH [expected OK] /html/semantics/forms/textfieldselection/selection-not-application.html
  │ 
  │ thread 'ScriptThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(0) }' panicked at 'DOMRefCell<T> already mutably borrowed', ../src/libcore/option.rs:704
  │ stack backtrace:
  │    1:        0x109d19e58 - sys::backtrace::tracing::imp::write::h3153bab3099b48114Zu
  │    2:        0x109d1da49 - panicking::default_handler::_$u7b$$u7b$closure$u7d$$u7d$::closure.43920
  │    3:        0x109d1d678 - panicking::default_handler::h1183b2a11c8682560zz
  │    4:        0x109d08a46 - sys_common::unwind::begin_unwind_inner::h1f97bc9e4cfe0396LZt
  │    5:        0x109d09a7e - sys_common::unwind::begin_unwind_fmt::hda8646f04aa67d8aRYt
  │    6:        0x109d191f7 - rust_begin_unwind
  │    7:        0x109d43650 - panicking::panic_fmt::h40f5ec0cdc3fc429FRL
  │    8:        0x109d4e935 - option::expect_failed::h56e104933d0889c7LpO
  │    9:        0x108707e7e - dom::htmlinputelement::HTMLInputElement.HTMLInputElementMethods::SetSelectionStart::hb411b730858a10c0HM3
  │   10:        0x108b02af0 - dom::bindings::codegen::Bindings::HTMLInputElementBinding::set_selectionStart::hc29799c26c5054eeDdo
  │   11:        0x1093df0aa - CallJitSetterOp
  │   12:        0x108ac6a5d - dom::bindings::utils::call_setter::h3621e47e57f90dd8Sug
  │   13:        0x108ac68df - dom::bindings::utils::generic_call::h2289bd9fb0388cfdcqg
  │   14:        0x10954b97b - _ZN2js6InvokeEP9JSContextN2JS8CallArgsENS_14MaybeConstructE
  │   15:        0x109543f7d - _ZN2js6InvokeEP9JSContextRKN2JS5ValueES5_jPS4_NS2_13MutableHandleIS3_EE
  │   16:        0x109561bae - _ZN2js20InvokeGetterOrSetterEP9JSContextP8JSObjectN2JS5ValueEjPS5_NS4_13MutableHandleIS5_EE
  │   17:        0x10957f937 - _ZN2js17NativeSetPropertyEP9JSContextN2JS6HandleIPNS_12NativeObjectEEENS3_IP8JSObjectEENS3_I4jsidEENS_13QualifiedBoolENS2_13MutableHandleINS2_5ValueEEERNS2_14ObjectOpResultE
  │   18:        0x10955aa95 - _ZL9InterpretP9JSContextRN2js8RunStateE
  │   19:        0x109552eb9 - _ZN2js9RunScriptEP9JSContextRNS_8RunStateE
  │   20:        0x10954bbb7 - _ZN2js6InvokeEP9JSContextN2JS8CallArgsENS_14MaybeConstructE
  │   21:        0x10987695c - _ZN2js9fun_applyEP9JSContextjPN2JS5ValueE
  │   22:        0x10954b97b - _ZN2js6InvokeEP9JSContextN2JS8CallArgsENS_14MaybeConstructE
  │   23:        0x109543f7d - _ZN2js6InvokeEP9JSContextRKN2JS5ValueES5_jPS4_NS2_13MutableHandleIS3_EE
  │   24:        0x109689fab - _ZN2js3jitL14DoCallFallbackEP9JSContextPNS0_13BaselineFrameEPNS0_15ICCall_FallbackEjPN2JS5ValueENS7_13MutableHandleIS8_EE
  └ fatal runtime error: Could not unwind stack, error = 5

  ▶ CRASH [expected OK] /html/semantics/forms/textfieldselection/selection-not-application-textarea.html
  │ 
  │ thread 'ScriptThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(0) }' panicked at 'DOMRefCell<T> already mutably borrowed', ../src/libcore/option.rs:704
  │ stack backtrace:
  │    1:        0x10e08ee58 - sys::backtrace::tracing::imp::write::h3153bab3099b48114Zu
  │    2:        0x10e092a49 - panicking::default_handler::_$u7b$$u7b$closure$u7d$$u7d$::closure.43920
  │    3:        0x10e092678 - panicking::default_handler::h1183b2a11c8682560zz
  │    4:        0x10e07da46 - sys_common::unwind::begin_unwind_inner::h1f97bc9e4cfe0396LZt
  │    5:        0x10e07ea7e - sys_common::unwind::begin_unwind_fmt::hda8646f04aa67d8aRYt
  │    6:        0x10e08e1f7 - rust_begin_unwind
  │    7:        0x10e0b8650 - panicking::panic_fmt::h40f5ec0cdc3fc429FRL
  │    8:        0x10e0c3935 - option::expect_failed::h56e104933d0889c7LpO
  │    9:        0x10ca8df3e - dom::htmltextareaelement::HTMLTextAreaElement.HTMLTextAreaElementMethods::SetSelectionStart::h531cfc2395a3da38yo8
  │   10:        0x10cc7d900 - dom::bindings::codegen::Bindings::HTMLTextAreaElementBinding::set_selectionStart::ha0d76266b710c2d4rKC
  │   11:        0x10d7540aa - CallJitSetterOp
  │   12:        0x10ce3ba5d - dom::bindings::utils::call_setter::h3621e47e57f90dd8Sug
  │   13:        0x10ce3b8df - dom::bindings::utils::generic_call::h2289bd9fb0388cfdcqg
  │   14:        0x10d8c097b - _ZN2js6InvokeEP9JSContextN2JS8CallArgsENS_14MaybeConstructE
  │   15:        0x10d8b8f7d - _ZN2js6InvokeEP9JSContextRKN2JS5ValueES5_jPS4_NS2_13MutableHandleIS3_EE
  │   16:        0x10d8d6bae - _ZN2js20InvokeGetterOrSetterEP9JSContextP8JSObjectN2JS5ValueEjPS5_NS4_13MutableHandleIS5_EE
  │   17:        0x10d8f4937 - _ZN2js17NativeSetPropertyEP9JSContextN2JS6HandleIPNS_12NativeObjectEEENS3_IP8JSObjectEENS3_I4jsidEENS_13QualifiedBoolENS2_13MutableHandleINS2_5ValueEEERNS2_14ObjectOpResultE
  │   18:        0x10d8cfa95 - _ZL9InterpretP9JSContextRN2js8RunStateE
  │   19:        0x10d8c7eb9 - _ZN2js9RunScriptEP9JSContextRNS_8RunStateE
  │   20:        0x10d8c0bb7 - _ZN2js6InvokeEP9JSContextN2JS8CallArgsENS_14MaybeConstructE
  │   21:        0x10dbeb95c - _ZN2js9fun_applyEP9JSContextjPN2JS5ValueE
  │   22:        0x10d8c097b - _ZN2js6InvokeEP9JSContextN2JS8CallArgsENS_14MaybeConstructE
  │   23:        0x10d8d14d9 - _ZL9InterpretP9JSContextRN2js8RunStateE
  │   24:        0x10d8c7eb9 - _ZN2js9RunScriptEP9JSContextRNS_8RunStateE
  │   25:        0x10d8d6d36 - _ZN2js13ExecuteKernelEP9JSContextN2JS6HandleIP8JSScriptEER8JSObjectRKNS2_5ValueENS_11ExecuteTypeENS_16AbstractFramePtrEPS9_
  │   26:        0x10d8d6e63 - _ZN2js7ExecuteEP9JSContextN2JS6HandleIP8JSScriptEER8JSObjectPNS2_5ValueE
  │   27:        0x10dbb6f02 - _ZL8EvaluateP9JSContextN2JS6HandleIP8JSObjectEERKNS1_22ReadOnlyCompileOptionsERNS1_18SourceBufferHolderENS1_13MutableHandleINS1_5ValueEEE
  │   28:        0x10dbb706a - _ZN2JS8EvaluateEP9JSContextRKNS_22ReadOnlyCompileOptionsEPKDsmNS_13MutableHandleINS_5ValueEEE
  │   29:        0x10ced25fa - dom::htmlscriptelement::HTMLScriptElement::execute::hd985d34a6a844843Sr6
  │   30:        0x10cefffa9 - dom::htmlscriptelement::HTMLScriptElement::prepare::he74a5354aaa7d643mh6
  │   31:        0x10caa1bb0 - tree_builder::rules::super..TreeBuilder<Handle, Sink>.TreeBuilderStep::step::h14576230894425074504
  │   32:        0x10ca9990f - tree_builder::TreeBuilder<Handle, Sink>.TokenSink::process_token::h7834235262675994692
  │   33:        0x10ca98ef1 - tokenizer::Tokenizer<Sink>::process_token::h6160948021956107981
  │   34:        0x10cac934a - tokenizer::Tokenizer<Sink>::emit_current_tag::h15594130976656807786
  │   35:        0x10cac36dd - tokenizer::Tokenizer<Sink>::step::h16952525606794884413
  │   36:        0x10cabe8c7 - tokenizer::Tokenizer<Sink>::run::h13044294324310735172
  │   37:        0x10ca93bee - dom::servohtmlparser::ServoHTMLParser::parse_sync::h9952bd596f81f8e2cSf
  │   38:        0x10cecf1bf - dom::document::Document::finish_load::h89ddd8470eb96f7eVLO
  │   39:        0x10cefe50d - dom::htmlscriptelement::ScriptContext.AsyncResponseListener::response_complete::h2b001ff6297d70d3Sf6
  │   40:        0x10d62bdda - ResponseAction::process::h69a97ad5099cd1a9Ypf
  │   41:        0x10cf01bb0 - network_listener::ListenerRunnable<T>.Runnable::handler::h5208449697483038179
  │   42:        0x10cda49bb - script_thread::ScriptThread::handle_msg_from_script::h0b0c9a560280f108Xzt
  │   43:        0x10ce0509a - script_thread::ScriptThread::handle_msgs::_$u7b$$u7b$closure$u7d$$u7d$::closure.188824
  │   44:        0x10cdedefe - script_thread::ScriptThread::handle_msgs::h45aab595807156d2Blt
  │   45:        0x10cd943c7 - sys_common::unwind::try::try_fn::h3279682059033863226
  │   46:        0x10e08e18b - __rust_try
  │   47:        0x10e08e113 - sys_common::unwind::inner_try::hdf6affff6fc30ecdNWt
  │   48:        0x10cd9581a - boxed::F.FnBox<A>::call_box::h15091847669718450595
  │   49:        0x10e091b2c - sys::thread::Thread::new::thread_start::he33a0e5f2597fcdejOy
  │   50:     0x7fff8face059 - _pthread_body
  └   51:     0x7fff8facdfd6 - _pthread_start

  ▶ CRASH [expected OK] /html/semantics/forms/the-input-element/selection.html
  │ 
  │ thread 'ScriptThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(0) }' panicked at 'DOMRefCell<T> already mutably borrowed', ../src/libcore/option.rs:704
  │ stack backtrace:
  │    1:        0x10edc6e58 - sys::backtrace::tracing::imp::write::h3153bab3099b48114Zu
  │    2:        0x10edcaa49 - panicking::default_handler::_$u7b$$u7b$closure$u7d$$u7d$::closure.43920
  │    3:        0x10edca678 - panicking::default_handler::h1183b2a11c8682560zz
  │    4:        0x10edb5a46 - sys_common::unwind::begin_unwind_inner::h1f97bc9e4cfe0396LZt
  │    5:        0x10edb6a7e - sys_common::unwind::begin_unwind_fmt::hda8646f04aa67d8aRYt
  │    6:        0x10edc61f7 - rust_begin_unwind
  │    7:        0x10edf0650 - panicking::panic_fmt::h40f5ec0cdc3fc429FRL
  │    8:        0x10edfb935 - option::expect_failed::h56e104933d0889c7LpO
  │    9:        0x10d7b4e7e - dom::htmlinputelement::HTMLInputElement.HTMLInputElementMethods::SetSelectionStart::hb411b730858a10c0HM3
  │   10:        0x10dbafaf0 - dom::bindings::codegen::Bindings::HTMLInputElementBinding::set_selectionStart::hc29799c26c5054eeDdo
  │   11:        0x10e48c0aa - CallJitSetterOp
  │   12:        0x10db73a5d - dom::bindings::utils::call_setter::h3621e47e57f90dd8Sug
  │   13:        0x10db738df - dom::bindings::utils::generic_call::h2289bd9fb0388cfdcqg
  │   14:        0x10e5f897b - _ZN2js6InvokeEP9JSContextN2JS8CallArgsENS_14MaybeConstructE
  │   15:        0x10e5f0f7d - _ZN2js6InvokeEP9JSContextRKN2JS5ValueES5_jPS4_NS2_13MutableHandleIS3_EE
  │   16:        0x10e60ebae - _ZN2js20InvokeGetterOrSetterEP9JSContextP8JSObjectN2JS5ValueEjPS5_NS4_13MutableHandleIS5_EE
  │   17:        0x10e62c937 - _ZN2js17NativeSetPropertyEP9JSContextN2JS6HandleIPNS_12NativeObjectEEENS3_IP8JSObjectEENS3_I4jsidEENS_13QualifiedBoolENS2_13MutableHandleINS2_5ValueEEERNS2_14ObjectOpResultE
  │   18:        0x10e607a95 - _ZL9InterpretP9JSContextRN2js8RunStateE
  │   19:        0x10e5ffeb9 - _ZN2js9RunScriptEP9JSContextRNS_8RunStateE
  │   20:        0x10e5f8bb7 - _ZN2js6InvokeEP9JSContextN2JS8CallArgsENS_14MaybeConstructE
  │   21:        0x10e92395c - _ZN2js9fun_applyEP9JSContextjPN2JS5ValueE
  │   22:        0x10e5f897b - _ZN2js6InvokeEP9JSContextN2JS8CallArgsENS_14MaybeConstructE
  │   23:        0x10e5f0f7d - _ZN2js6InvokeEP9JSContextRKN2JS5ValueES5_jPS4_NS2_13MutableHandleIS3_EE
  │   24:        0x10e736fab - _ZN2js3jitL14DoCallFallbackEP9JSContextPNS0_13BaselineFrameEPNS0_15ICCall_FallbackEjPN2JS5ValueENS7_13MutableHandleIS8_EE
  └ fatal runtime error: Could not unwind stack, error = 5
```

bors-servo commented:
:umbrella: The latest upstream changes (presumably #10231) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
Closing due to inactivity. Thanks for the work you accomplished, @0X1A!
